docs/resources/server/recipes/custom.md

```diff
@@ -0,0 +1,699 @@
+        DeferredRegister.create(Registries.RECIPE_SERIALIZER, ExampleMod.MOD_ID);
+
+public static final Supplier<RecipeSerializer<RightClickBlockRecipe>> RIGHT_CLICK_BLOCK =
+        RECIPE_SERIALIZERS.register("right_click_block", RightClickBlockRecipeSerializer::new);
+```
+
+And similarly, we must also override `#getSerializer` in our recipe, like so:
+
+```java
+public class RightClickBlockRecipe implements Recipe<RightClickBlockInput> {
+    // other stuff here
+
+    @Override
+    public RecipeSerializer<? extends Recipe<RightClickBlockInput>> getSerializer() {
+        return RIGHT_CLICK_BLOCK.get();
+    }
+}
+```
+
+## The Crafting Mechanic
+
+Now that all parts of your recipe are complete, you can make yourself some recipe JSONs (see the [datagen] section for that) and then query the recipe manager for your recipes, like above. What you then do with the recipe is up to you. A common use case would be a machine that can process your recipes, storing the active recipe as a field.
+
+In our case, however, we want to apply the recipe when an item is right-clicked on a block. We will do so using an [event handler][event]. Keep in mind that this is an example implementation, and you can alter this in any way you like (so long as you run it on the server). As we want the interaction state to match on both the client and server, we will also need to [sync any relevant input states across the network][networking].
+
+We can set up a simple network implementation to sync the recipe inputs like so:
+
+```java
+// A basic packet class, must be registered.
+public record ClientboundRightClickBlockRecipesPayload(
+    Set<BlockState> inputStates, Set<Holder<Item>> inputItems
+) implements CustomPacketPayload {
+
+    // ...
+}
+
+
+// Packet stores data in an instance class.
+// Present on both server and client to do initial matching.
+// Resource listener so it can be reloaded when recipes are.
+public class RightClickBlockRecipeInputs extends SimplePreparableReloadListener<Void> {
+    // Only one instance
+    public static final RightClickBlockRecipeInputs INSTANCE = new RightClickBlockRecipeInputs();
+
+    private Set<BlockState> inputStates;
+    private Set<Holder<Item>> inputItems;
+
+    private RightClickBlockRecipeInputs() {}
+
+    @Override
+    protected Void prepare(ResourceManager manager, ProfilerFiller filler) {}
+
+    // Set inputs here as #apply is fired synchronously based on listener registration order.
+    // Recipes are always applied first.
+    @Override
+    protected abstract void apply(Void dummy, ResourceManager manager, ProfilerFiller filler) {
+        MinecraftServer server = ServerLifecycleHooks.getCurrentServer();
+        if (server != null) { // Should never be null
+            // Populate inputs
+            Set<BlockState> inputStates = new HashSet<>();
+            Set<Holder<Item>> inputItems = new HashSet<>();
+
+            server.getRecipeManager().recipeMap().byType(RIGHT_CLICK_BLOCK_TYPE.get())
```

Comment:
The `RecipeManager` should be acquired by capturing the `ReloadableServerResources` in the `AddReloadListenerEvent` instead of going through the server. This of course requires the "reloader" to be a separate class from the rest of this recipe input cache.